### PR TITLE
Add etcd defragmentation service + timer.

### PR DESCRIPTION
**Pull request by garloff**
Unconditionally apply heartbeat/election timing tweaks. Unnice/IO-unnice etcd.
Wait for etcd to have restarted before doing it.

More importantly: Add nightly etcd-defrag job. (#279) Heavily inspired by Sven Batista Steinbach from syself' work on the Cluster-API based Cluster Management for Hetzner.

Signed-off-by: Kurt Garloff <kurt@garloff.de>

### CODE REVIEW

**Comment by garloff** [comment]
It does not quite work yet ...
```
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): Optional Variables:                                                     
openstack_compute_instance_v2.mgmtcluster_server (remote-exec):   - CLUSTER_NAME                 (defaults to tc1)                      
openstack_compute_instance_v2.mgmtcluster_server (remote-exec):   - CONTROL_PLANE_MACHINE_COUNT  (defaults to 1)                        
openstack_compute_instance_v2.mgmtcluster_server (remote-exec):   - WORKER_MACHINE_COUNT         (defaults to 0)                        
                                                                                                                                        
openstack_compute_instance_v2.mgmtcluster_server (remote-exec):                                                                         
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): New clusterctl version available: v1.2.1 -> v1.2.2                      
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.2.2      
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): # rendering clusterconfig from template                                 
openstack_compute_instance_v2.mgmtcluster_server: Still creating... [6m0s elapsed]                                                      openstack_compute_instance_v2.mgmtcluster_server (remote-exec):                                                                         
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): New clusterctl version available: v1.2.1 -> v1.2.2                      
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): https://github.com/kubernetes-sigs/cluster-api/releases/tag/v1.2.2      
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): # apply configuration and deploy cluster tc1                            
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): cluster.cluster.x-k8s.io/tc1 created                                    
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): openstackcluster.infrastructure.cluster.x-k8s.io/tc1 created            
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): kubeadmcontrolplane.controlplane.cluster.x-k8s.io/tc1-control-plane created                                                                                                                                     openstack_compute_instance_v2.mgmtcluster_server (remote-exec): openstackmachinetemplate.infrastructure.cluster.x-k8s.io/capi1-tc1-control-plane-genc01 created                                                                                                                 
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): machinedeployment.cluster.x-k8s.io/capi1-tc1-md-0-no1 created           
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): openstackmachinetemplate.infrastructure.cluster.x-k8s.io/capi1-tc1-md-0-genw01 created                                                                                                                          
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): secret/tc1-cloud-config created                                         
openstack_compute_instance_v2.mgmtcluster_server (remote-exec): Error from server (InternalError): error when creating "/home/ubuntu/tc1/tc1-config.yaml": Internal error occurred: failed calling webhook "default.kubeadmconfigtemplate.bootstrap.cluster.x-k8s.io": failed to call webhook: Post "https://capi-kubeadm-bootstrap-webhook-service.capi-kubeadm-bootstrap-system.svc:443/mutate-bootstrap-cluster-x-k8s-io-v1beta1-kubeadmconfigtemplate?timeout=10s": dial tcp 10.96.116.37:443: connect: connection refused
```

**Comment by batistein** [comment]
I would put the files for systemd timer under files. Then it's human readable. See: https://github.com/kubernetes-sigs/cluster-api/blob/main/bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go#L52-L54 
or:
https://github.com/kubernetes-sigs/cluster-api/blob/main/controlplane/kubeadm/config/crd/bases/controlplane.cluster.x-k8s.io_kubeadmcontrolplanes.yaml#L2875

`dial tcp 10.96.116.37:443: connect: connection refused` This sounds like the webhook server of the bootstrap provider kubeadm is not ready yet.

**Comment by batistein** [comment]
example how to use files: https://github.com/syself/cluster-api-provider-hetzner/blob/main/templates/cluster-templates/bases/hcloud-kcp-ubuntu.yaml#L136-L154

**Comment by garloff** [comment]
Thanks for your hints, @batistein , very much appreciated.
It looks like capo can end up being seriously in trouble:
```
I0919 06:47:44.572133       1 openstackmachine_controller.go:202] controller/openstackmachine "msg"="Reconciling Machine delete" "cluster"="tc1" "machine"="capi1-tc1-md-0-no1-6758b8f59c-wkzw5" "name"="capi1-tc1-md-0-genw01-2gw5j" "namespace"="default" "openStackCluster"="tc1" "openStackMachine"="capi1-tc1-md-0-genw01-2gw5j" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="OpenStackMachine" 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x14e98c6]

goroutine 396 [running]:
sigs.k8s.io/cluster-api-provider-openstack/controllers.machineToInstanceSpec(0xc0006a9c00, 0xc00077f8c0, 0xc0000b1680, {0x0, 0x0})
        /workspace/controllers/openstackmachine_controller.go:474 +0x546
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).reconcileDelete(0x1a6c338, {0x1a6c338, 0xc000713f20}, 0xc0003c5380, 0xc0000b1680, 0xc0003f6e00, 0xc0006a9c00, 0xc00077f8c0, 0xc0000b1680)
        /workspace/controllers/openstackmachine_controller.go:252 +0x46b
sigs.k8s.io/cluster-api-provider-openstack/controllers.(*OpenStackMachineReconciler).Reconcile(0xc000631cb0, {0x1a6c338, 0xc000713f20}, {{{0xc0006bcbd6, 0x1766dc0}, {0xc0004cb140, 0x30}}})
        /workspace/controllers/openstackmachine_controller.go:157 +0xb05
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0xc0000b8370, {0x1a6c338, 0xc000713ef0}, {{{0xc0006bcbd6, 0x1766dc0}, {0xc0004cb140, 0x413c34}}})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:114 +0x26f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0000b8370, {0x1a6c290, 0xc000658040}, {0x16a4c00, 0xc00071a940})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:311 +0x33e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0000b8370, {0x1a6c290, 0xc000658040})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:223 +0x357
```
Not the kind of things I want to find in logs.

**Comment by batistein** [comment]
@garloff did you create a worker security group? See: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/v0.6.3/controllers/openstackmachine_controller.go#L474

Nevertheless this shouldn't lead to such an issue...

**Comment by garloff** [comment]
OK, this works nicely now.
etcd-defrag unit files deployed via the file mechanism, etcd gets its parameters via kubeadmconfig.
Ready for review :-)

File: terraform/files/template/cluster-template.yaml [hunk 1]
```diff
@@ -63,15 +63,64 @@ spec:
...
         kubeletExtraArgs:
           cloud-provider: external
+    files:
+      - path: /etc/systemd/system/etcd-defrag.service
+        owner: "root:root"
+        permissions: "0644"
+        content: |
+          [Unit]
+          Description=Run etcdctl defrag
+          Documentation=https://etcd.io/docs/v3.3.12/op-guide/maintenance/#defragmentation
```
Comment (batistein) [inline terraform/files/template/cluster-template.yaml:87]: we should update this to the current used version (latest: https://github.com/etcd-io/etcd/releases)

File: terraform/files/template/cluster-template.yaml [hunk 2]
```diff
@@ -63,15 +63,64 @@ spec:
...
+          Type=oneshot
+          Environment="LOG_DIR=/var/log"
+          Environment="ETCDCTL_API=3"
+          ExecStart=sync
+          ExecStart=sleep 2
+          ExecStart=etcdctl --cert /etc/kubernetes/pki/etcd/peer.crt --key /etc/kubernetes/pki/etcd/peer.key --cacert /etc/kubernetes/pki/etcd/ca.crt defrag
+          ExecStart=sleep 3
+          ExecStart=etcdctl --cert /etc/kubernetes/pki/etcd/peer.crt --key /etc/kubernetes/pki/etcd/peer.key --cacert /etc/kubernetes/pki/etcd/ca.crt snapshot save /root/etcd-backup
+          ExecStart=xz -f /root/etcd-backup
+          ExecStart=fstrim -v /var/lib/etcd
```
Comment (batistein) [inline terraform/files/template/cluster-template.yaml:100]: What does this do? I'm not familar with that tool...

**Review by batistein — APPROVED** [review]

**Pull request merged**